crates/bcr-wallet-core/src/purse.rs

```diff
@@ -9,31 +10,77 @@ use crate::{error::Result, types::WalletConfig};
 #[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
 #[cfg_attr(not(target_arch = "wasm32"), async_trait)]
 pub trait PurseRepository {
-    async fn store_wallet(&self, wallet: WalletConfig) -> Result<()>;
-    async fn load_wallet(&self, wallet_id: &str) -> Result<WalletConfig>;
-    async fn delete_wallet(&self, wallet_id: &str) -> Result<()>;
-    async fn list_wallets(&self) -> Result<Vec<String>>;
+    async fn store(&self, wallet: WalletConfig) -> Result<()>;
+    async fn load(&self, wallet_id: &str) -> Result<WalletConfig>;
+    async fn delete(&self, wallet_id: &str) -> Result<()>;
+    async fn list_ids(&self) -> Result<Vec<String>>;
 }
 
-pub struct Purse<PurseRepo> {
-    pub wallets: PurseRepo,
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+pub trait Wallet {
+    fn config(&self) -> WalletConfig;
+    fn name(&self) -> String;
+}
+
+pub struct Purse<PurseRepo, Wlt> {
+    pub repo: PurseRepo,
+    pub wallets: Mutex<Vec<Arc<Wlt>>>,
+}
+impl<PurseRepo, Wlt> Purse<PurseRepo, Wlt> {
+    pub fn new(repo: PurseRepo) -> Self {
+        Self {
+            repo,
+            wallets: Mutex::new(Vec::default()),
+        }
+    }
 }
 
-impl<PurseRepo> Purse<PurseRepo>
+impl<PurseRepo, Wlt> Purse<PurseRepo, Wlt>
 where
     PurseRepo: PurseRepository,
 {
-    pub async fn store_wallet(&self, wallet: WalletConfig) -> Result<()> {
-        self.wallets.store_wallet(wallet).await
+    pub async fn load_wallet_config(&self, wallet_id: &str) -> Result<WalletConfig> {
+        self.repo.load(wallet_id).await
     }
-    pub async fn load_wallet(&self, wallet_id: &str) -> Result<WalletConfig> {
-        self.wallets.load_wallet(wallet_id).await
+    pub async fn list_wallets(&self) -> Result<Vec<String>> {
+        self.repo.list_ids().await
     }
-    #[allow(dead_code)]
-    pub async fn delete_wallet(&self, wallet_id: &str) -> Result<()> {
-        self.wallets.delete_wallet(wallet_id).await
+
+    pub fn get_wallet(&self, idx: usize) -> Option<Arc<Wlt>> {
+        let wallets = self.wallets.lock().unwrap();
+        if idx < wallets.len() {
```

Comment:
Wouldn't `wallets.get(idx)` maybe with a `.map(|w| w.clone())` or `.cloned()` be essentially the same? 🤔